lib/telemetry/armon_metrics.go

```diff
@@ -327,57 +401,92 @@ func circonusSink(cfg TelemetryConfig, hostname string) (metrics.MetricSink, err
 	return sink, nil
 }
 
-// InitTelemetry configures go-metrics based on map of telemetry config
-// values as returned by Runtimecfg.Config().
-func InitTelemetry(cfg TelemetryConfig) (*metrics.InmemSink, error) {
+// addSink takes our collection of sinks and applies configuration to its setup function, appending the new sink to the
+// collection.
+func addSink(sinks metrics.FanoutSink, fn sinkFn, cfg Config, hostname string) error {
+	// Build the sink
+	s, err := fn(cfg, hostname)
+	if err != nil {
+		return err
+	}
+	// Compose into FanoutSink
+	if s != nil {
+		sinks = append(sinks, s)
+	}
```

Comment:
I suspect this is not working because the `sinks` slice is not being returned. The backing array is being populated, but we still need the slice to know how long it is at the end. Arrays are mutable, but append does not modify the passed in slice, it creates a new one.

I think you could keep this `addSink` as an nested function in `initSinks` to avoid that problem:

```go

type sinkFn func(Config) (metrics.MetricSink, error)
...

addSink := func(fn sinkFn, cfg Config) error {
	s, err := fn(cfg)
	if err != nil || s == nil {
		return err
	}
	sinks = append(sinks, s)
	return nil
}


```

Reply:
Yeah, I'm not crazy about how these functions reach up to modify a variable in the scope above it. I started changing it to take and return args but lost track of it during an earlier pass, leaving a bug. Thanks for spotting this! I've reverted it.